src/traefik.py

```diff
@@ -0,0 +1,600 @@
+            }
+        ]
+
+    def _update_tls_configuration(self):
+        """Generate and push tls config yaml for traefik."""
+        config = yaml.safe_dump(
+            {
+                "tls": {
+                    "certificates": [
+                        {
+                            "certFile": SERVER_CERT_PATH,
+                            "keyFile": SERVER_KEY_PATH,
+                        }
+                    ],
+                    "stores": {
+                        "default": {
+                            # When the external hostname is a bare IP, traefik cannot match a domain,
+                            # so we must set the default cert for the TLS handshake to succeed.
+                            "defaultCertificate": {
+                                "certFile": SERVER_CERT_PATH,
+                                "keyFile": SERVER_KEY_PATH,
+                            },
+                        },
+                    },
+                }
+            }
+        )
+        self._container.push(DYNAMIC_CERTS_PATH, config, make_dirs=True)
+
+    def configure(self):
+        """Configure static and tls."""
+        # Ensure the required basic configurations and folders exist
+        tcp_entrypoints = self._tcp_entrypoints
+        self._update_static_configuration(tcp_entrypoints)
+        self._setup_dynamic_config_folder()
+
+        if self._tls_enabled:
+            self._update_tls_configuration()
+
+    def update_cert_configuration(
+        self, cert: Optional[str], key: Optional[str], ca: Optional[str]
+    ):
+        """Update the server cert, ca, and key configuration files."""
+        if cert:
+            self._container.push(SERVER_CERT_PATH, cert, make_dirs=True)
+        else:
+            self._container.remove_path(SERVER_CERT_PATH, recursive=True)
+
+        if key:
+            self._container.push(SERVER_KEY_PATH, key, make_dirs=True)
+        else:
+            self._container.remove_path(SERVER_KEY_PATH, recursive=True)
+
+        if ca:
+            self._container.push(CA_CERT_PATH, ca, make_dirs=True)
+        else:
+            self._container.remove_path(CA_CERT_PATH, recursive=True)
+
+        self.update_ca_certs()
+
+    def add_ca(self, ca: str, uid: Union[str, int]):
+        """Add a ca."""
+        ca_path = RECV_CA_TEMPLATE.substitute(rel_id=str(uid))
+        self._container.push(ca_path, ca, make_dirs=True)
+
+    def remove_ca(self, uid: Union[str, int]):
+        """Remove a ca.
+
+        BEWARE of potential race conditions.
+        Traefik watches the dynamic config dir and reloads automatically on change.
+        So make sure any traefik config depending on the certificates being there is updated
+        **before** the certs are removed, otherwise you might have some downtime.
+        """
+        ca_path = RECV_CA_TEMPLATE.substitute(rel_id=str(uid))
+        self._container.remove_path(ca_path)
```

Comment:
same as above: 
![image](https://github.com/canonical/traefik-k8s-operator/assets/6230162/13d658c6-1f42-4dda-83f0-9fc0e162e5bf)

Although we're not really doing any 'nested loop' ca removals, so we could simplify this flow by making `remove_ca` always automatically call update-ca-certificates, but I think this way it's more symmetric.

Another option is to refactor `add_ca` and `remove_ca` to:
`add_cas(List[CA])`, and make this method call always invoke automatically update-ca-certificates after adding all the CAs. Analogously for remove_ca --> remove_cas.

this would mean the charm is no longer responsible for calling update_ca_certificates after it's done adding/removing the CAs.

Reply:
in light of more comments below, I went for the refactoring mentioned ^